library/core/src/pin.rs

```diff
@@ -195,181 +628,273 @@
 //! meaning that if the struct is pinned then so is the field.
 //!
-//! This allows writing a projection that creates a <code>[Pin]<[&mut] Field></code>, thus
+//! This allows writing a projection that creates a <code>[Pin]<[`&mut Field`]></code>, thus
 //! witnessing that the field is pinned:
 //!
 //! ```rust,no_run
 //! # use std::pin::Pin;
 //! # type Field = i32;
 //! # struct Struct { field: Field }
 //! impl Struct {
-//!     fn pin_get_field(self: Pin<&mut Self>) -> Pin<&mut Field> {
+//!     fn field(self: Pin<&mut Self>) -> Pin<&mut Field> {
 //!         // This is okay because `field` is pinned when `self` is.
 //!         unsafe { self.map_unchecked_mut(|s| &mut s.field) }
 //!     }
 //! }
 //! ```
 //!
-//! However, structural pinning comes with a few extra requirements:
-//!
-//! 1.  The struct must only be [`Unpin`] if all the structural fields are
-//!     [`Unpin`]. This is the default, but [`Unpin`] is a safe trait, so as the author of
-//!     the struct it is your responsibility *not* to add something like
-//!     <code>impl\<T> [Unpin] for Struct\<T></code>. (Notice that adding a projection operation
-//!     requires unsafe code, so the fact that [`Unpin`] is a safe trait does not break
-//!     the principle that you only have to worry about any of this if you use [`unsafe`].)
-//! 2.  The destructor of the struct must not move structural fields out of its argument. This
-//!     is the exact point that was raised in the [previous section][drop-impl]: [`drop`] takes
-//!     <code>[&mut] self</code>, but the struct (and hence its fields) might have been pinned
-//!     before. You have to guarantee that you do not move a field inside your [`Drop`][Drop]
-//!     implementation. In particular, as explained previously, this means that your struct
-//!     must *not* be `#[repr(packed)]`.
-//!     See that section for how to write [`drop`] in a way that the compiler can help you
-//!     not accidentally break pinning.
-//! 3.  You must make sure that you uphold the [`Drop` guarantee][drop-guarantee]:
-//!     once your struct is pinned, the memory that contains the
-//!     content is not overwritten or deallocated without calling the content's destructors.
-//!     This can be tricky, as witnessed by <code>[VecDeque]\<T></code>: the destructor of
-//!     <code>[VecDeque]\<T></code> can fail to call [`drop`] on all elements if one of the
-//!     destructors panics. This violates the [`Drop`][Drop] guarantee, because it can lead to
-//!     elements being deallocated without their destructor being called.
-//!     (<code>[VecDeque]\<T></code> has no pinning projections, so this
-//!     does not cause unsoundness.)
-//! 4.  You must not offer any other operations that could lead to data being moved out of
+//! Structural pinning comes with a few extra requirements:
+//!
+//! 1.  *Structural [`Unpin`].* A struct can be [`Unpin`] if, and only if, all of its
```

Comment:
This is not quite an "only if" -- the struct itself could be "where the pinning starts", like a manual implementation of a self-referential future.

Reply:
This seems footnote-worthy but I don't think I can craft an appropriately nitpicky footnote for this

Reply:
I think I'd prefer some sort of rewording here. This is basically stating a definition of `Unpin` but it's a different definition from what other parts of the docs say. Really a struct can be `Unpin` if and only if it is "pinning-agnostic", to use the terms from the trait dos.

Reply:
I think this is good for a followup if you want to take a stab at it.